### DISCUSSION TITLE
Confused by dimension order in MONAI.

### DISCUSSION THREAD

**Discussion by GYDDHPY** [Project-MONAI/MONAI#4872] [Q&A] [answered]
I was reading SWIN UNETR code and was confused by the dimension order used in the model.  

For BTCV dataset, the preprocessing method used in SWIN UNETR are similiar to UNETR ,which include:    
`LoadImaged,AddChanneld,Orientationd,Spacingd,ScaleIntensityRanged,CropForegroundd,RandCropByPosNegLabeld,RandFlipd,RandFlipd,RandFlipd,RandRotate90d,RandScaleIntensityd,RandShiftIntensityd,ToTensord`.

The dimension order of the image after using these transforms seems to be `[h w d]`.  

However, in the `PatchEmbed` module of `SwinTransformer`, the code directly used `_, _, d, h, w = x_shape`, which treat the image as the dimension order changed to be `[d h w]`.
https://github.com/Project-MONAI/MONAI/blob/456caedcf2de5458834d2a10aa8ccca7959c61f6/monai/networks/blocks/patchembedding.py#L122-L180  

This confused me a lot. Is there any procedure changed the dimension order that I did't notice or something else?

**Comment by wyli** [comment, marked as answer]
In my understanding as long as the model is consistently used in training and inference, there's no need to distinguish the first/second/third spatial dimensions, these d, h, w are just variable names. the only problem I can see is that if after some training, the model is used together with a different preprocessing pipeline built outside of monai. in this case the pipeline should be created carefully so that it consistently reproduces the training ones.

**Reply by GYDDHPY** [reply]
Thank you for your quick reply. I'm trying to understand your answer.

The preprocessing method is designed for specific problem, for example, the patch size of each dimension is set for different axis. But the further variable order, such as d h w, is not required to be strictly consistent with the image axis.  Because the complicate neural network will deel with such design. All we need to be careful is to make the procedure of training and inference to be consistent. 

Is my understanding correct?  

By the way, does MONAI also have this in mind during the design process?

**Reply by wyli** [reply]
correct, and yes, the metadata dictionary and [more recently metatensor](https://github.com/Project-MONAI/MONAI/discussions/4764) are trying to capture the additional information associated with the image arrays such as spacing and orientation, then with the preprocessing transforms such as `Spacing` and `Orientation` will normalize the input data in a consistent manner. if these preprocessing are used for training, then the model weights become conditioned on these steps. one exception is the anisotropic models (with 3D conv kernels such as 3x3x1), for example https://github.com/Project-MONAI/MONAI/blob/456caedcf2de5458834d2a10aa8ccca7959c61f6/monai/networks/nets/ahnet.py#L50 it typically assumes 2d thick slices as input where the first two spatial dimensions have high resolution and the third dimension has large spacing.

**Reply by GYDDHPY** [reply]
Thank you for your quick reply again.  

Your answer helps me better understand the design of MONAI project.
